src/main/java/edu/kit/compiler/optimizations/common_subexpression/CommonSubexpressionElimination.java

```diff
@@ -0,0 +1,466 @@
+import firm.bindings.binding_irnode;
+import firm.nodes.*;
+import lombok.Data;
+import lombok.RequiredArgsConstructor;
+
+import java.util.*;
+
+import static firm.bindings.binding_irgraph.ir_graph_properties_t.IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE;
+
+@RequiredArgsConstructor
+@Data
+public class CommonSubexpressionElimination implements Optimization.Local {
+
+    // map that holds nodes that are to be exchanged with equivalent other nodes
+    private final HashMap<Node, Node> nodeValues = new HashMap<>();
+    private final Map<Pointer, Map<Pointer, Integer>> dominateMap = new HashMap<>();
+
+    @Override
+    public boolean optimize(Graph g, OptimizationState state) {
+        g.assureProperties(IR_GRAPH_PROPERTY_CONSISTENT_DOMINANCE);
+
+        // compare new change map with old one
+        boolean backEdgesEnabled = BackEdges.enabled(g);
+        if (!backEdgesEnabled) {
+            BackEdges.enable(g);
+        }
+
+        boolean hadChange = false;
+        boolean changes;
+        int maxChanges = 0;
+
+        do {
+            changes = false;
+            nodeValues.clear();
+            dominateMap.clear();
+
+            CSEVisitor visitor = new CSEVisitor(state);
+
+            List<Node> nodes = new ArrayList<>();
+            Util.NodeListFiller nodeCollector = new Util.NodeListFiller(nodes);
+            g.walkTopological(nodeCollector);
+            while (!nodes.isEmpty()) {
+                Node n = nodes.remove(0);
+                n.accept(visitor);
+            }
+
+            HashMap<Node, Node> replacementMap = new HashMap<>();
+            for (var replacement : nodeValues.entrySet()) {
+                // for each replacement merge nodes together
+                Node orig = replacement.getKey();
+                Node n = replacement.getValue();
+
+                int maxIndirections = 150;
+                while (orig.getOpCode() == binding_irnode.ir_opcode.iro_Deleted && maxIndirections > 0) {
+                    orig = replacementMap.getOrDefault(orig, orig);
+                    maxIndirections--;
+                }
+
+                while (n.getOpCode() == binding_irnode.ir_opcode.iro_Deleted && maxIndirections > 0) {
+                    n = replacementMap.getOrDefault(n, n);
+                    maxIndirections--;
+                }
+
+                if (maxIndirections == 0) {
+                    continue;
+                }
+
+                if (orig.getOpCode() != n.getOpCode()) {
+                    throw new RuntimeException("wrrong");
+                }
+                changes |= transform(g, orig, n, orig.getOpCode(), replacementMap);
+            }
+            if (changes) {
+                hadChange = true;
+                maxChanges++;
+            }
+
+            binding_irgopt.remove_bads(g.ptr);
+            binding_irgopt.remove_unreachable_code(g.ptr);
+            binding_irgopt.remove_bads(g.ptr);
+        } while (changes && maxChanges < 10);
```

Comment:
Is it necessary to have a fix point iteration here where a new `CSEVisitor` is created in every iteration?